File: src/Goal.Domain/Services/BlockingState.cs
namespace Goal.Domain.Services;

/// <summary>
/// The derived blocking state for a member in a sprint. Computed on demand (never persisted)
/// so it can never drift from the underlying XP/target values.
/// </summary>
public readonly record struct BlockingState
{
    public bool IsBlocked { get; init; }
    public int EarnedXp { get; init; }
    public int EffectiveTargetXp { get; init; }
    public int TargetXp { get; init; }              // the XP needed right now to be unblocked
    public int UnblockThresholdXp { get; init; }    // the normal (non-final) threshold
    public int DebtXp { get; init; }
    public int DaysRemaining { get; init; }
    public bool RequiresFullCompletion { get; init; }

    /// <summary>0..1 progress toward the effective target (capped at 1).</summary>
    public decimal CurrentPct =>
        EffectiveTargetXp <= 0 ? 1m : Math.Min(1m, (decimal)EarnedXp / EffectiveTargetXp);

    /// <summary>0..1 fraction of the effective target that the current target represents.</summary>
    public decimal TargetPct =>
        EffectiveTargetXp <= 0 ? 1m : Math.Min(1m, (decimal)TargetXp / EffectiveTargetXp);

    /// <summary>XP still needed to become unblocked (0 if already unblocked).</summary>
    public int XpRemainingToUnblock => Math.Max(0, TargetXp - EarnedXp);
}

File: app/lib/features/blocking/domain/blocking_state.dart
import '../../goals/domain/goal_models.dart';

/// Mirrors the backend BlockingStateDto. The app reflects this; it never decides blocking itself.
class BlockingState {
  const BlockingState({
    required this.isBlocked,
    required this.currentPct,
    required this.targetPct,
    required this.earnedXp,
    required this.effectiveTargetXp,
    required this.targetXp,
    required this.unblockThresholdXp,
    required this.debtXp,
    required this.daysRemaining,
    required this.requiresFullCompletion,
    required this.xpRemainingToUnblock,
    required this.blockedApps,
    this.randomOverlayActive = false,
    this.typingSabotageActive = false,
    this.typingSabotageText,
  });

  final bool isBlocked;
  final double currentPct;
  final double targetPct;
  final int earnedXp;
  final int effectiveTargetXp;
  final int targetXp;
  final int unblockThresholdXp;
  final int debtXp;
  final int daysRemaining;
  final bool requiresFullCompletion;
  final int xpRemainingToUnblock;
  final List<BlockedApp> blockedApps;
  final bool randomOverlayActive;
  final bool typingSabotageActive;
  final String? typingSabotageText;

  factory BlockingState.fromJson(Map<String, dynamic> j) => BlockingState(
        isBlocked: j['isBlocked'],
        currentPct: (j['currentPct'] as num).toDouble(),
        targetPct: (j['targetPct'] as num).toDouble(),
        earnedXp: j['earnedXp'],
        effectiveTargetXp: j['effectiveTargetXp'],
        targetXp: j['targetXp'],
        unblockThresholdXp: j['unblockThresholdXp'],
        debtXp: j['debtXp'],
        daysRemaining: j['daysRemaining'],
        requiresFullCompletion: j['requiresFullCompletion'],
        xpRemainingToUnblock: j['xpRemainingToUnblock'],
        blockedApps: (j['blockedApps'] as List).map((e) => BlockedApp.fromJson(e)).toList(),
        randomOverlayActive: (j['randomOverlayActive'] as bool?) ?? false,
        typingSabotageActive: (j['typingSabotageActive'] as bool?) ?? false,
        typingSabotageText: j['typingSabotageText'] as String?,
      );
}
